parser/parse.go

```diff
@@ -260,6 +260,18 @@ func (b *Builder) AddDirTo(dir string, u *types.Universe) error {
 	return b.findTypesIn(canonicalizeImportPath(b.buildPackages[dir].ImportPath), u)
 }
 
+// AddDirToAndReturnAddedPath is the same as AddDirTo except that it also
+// returns the path of the added package.
+func (b *Builder) AddDirToAndReturnAddedPath(dir string, u *types.Universe) (string, error) {
```

Comment:
Instead of doing this, add a method that canonicalizes a path and returns a *Package object.